File: collision_deformer/collision_deformer.py
import maya.OpenMaya as om
import maya.OpenMayaMPx as ommpx

class CollisionDeformer(ommpx.MPxDeformerNode):
    kNodeName = "collisionDeformer"
    kNodeClassify = "deformer"
    kNodeID = om.MTypeId(0x100000) # You should use a unique id for your node

    aBounciness = om.MObject()
    aFriction = om.MObject()

    def __init__(self):
        ommpx.MPxDeformerNode.__init__(self)

    def deform(self, data, itGeo, localToWorldMatrix, geomIndex):
        input = ommpx.cvar.MPxDeformerNode_input
        env = ommpx.cvar.MPxDeformerNode_envelope
        bounciness = data.inputValue(CollisionDeformer.aBounciness).asFloat()
        friction = data.inputValue(CollisionDeformer.aFriction).asFloat()

        while not itGeo.isDone():
            index = itGeo.index()
            inputGeom = ommpx.MFnMesh(input)
            point = inputGeom.getPoint(index, om.MSpace.kObject)
            
            # perform collision calculations here
            # use bounciness and friction attributes to control the behavior
            # of the collision

            inputGeom.setPoint(index, point, om.MSpace.kObject)

            itGeo.next()
            
    def accessoryNodeSetup(self, dagMod):
        pass

    def accessoryNodeConnections(self):
        pass

def nodeCreator():
    return ommpx.asMPxPtr(CollisionDeformer())

def nodeInitializer():
    nAttr = om.MFnNumericAttribute()

    CollisionDeformer.aBounciness = nAttr.create("bounciness", "b", om.MFnNumericData.kFloat, 1.0)
    nAttr.setKeyable(True)
    CollisionDeformer.addAttribute(CollisionDeformer.aBounciness)
    CollisionDeformer.attributeAffects(CollisionDeformer.aBounciness, ommpx.cvar.MPxDeformerNode_outputGeom)

    CollisionDeformer.aFriction = nAttr.create("friction", "f", om.MFnNumericData.kFloat, 0.5)
    nAttr.setKeyable(True)
    CollisionDeformer.addAttribute(CollisionDeformer.aFriction)
    CollisionDeformer.attributeAffects(CollisionDeformer.aFriction, ommpx.cvar.MPxDeformerNode_outputGeom)
    
def initializePlugin(mobject):
    mplugin = om.MFnPlugin(mobject)
    try:
        mplugin.registerNode(CollisionDeformer.kNodeName, CollisionDeformer.kNodeID, nodeCreator, nodeInitializer, ommpx.MPxNode.kDeformerNode)
    except:
        sys.stderr.write("Failed to register node: %s" % CollisionDeformer.kNodeName)
        raise

File: collision_deformer/cpp/collision_deformer.cpp
#include <maya/MFnNumericAttribute.h>
#include <maya/MFnMesh.h>
#include <maya/MItGeometry.h>
#include <maya/MPxDeformerNode.h>

class CollisionDeformer : public MPxDeformerNode
{
public:
    CollisionDeformer();
    virtual ~CollisionDeformer();

    static void* creator();
    static MStatus initialize();

    virtual MStatus deform(MDataBlock& data, MItGeometry& itGeo, const MMatrix& localToWorldMatrix, unsigned int geomIndex);

    static MTypeId id;
    static MObject aBounciness;
    static MObject aFriction;
};

MTypeId CollisionDeformer::id(0x100000);
MObject CollisionDeformer::aBounciness;
MObject CollisionDeformer::aFriction;

CollisionDeformer::CollisionDeformer() {}
CollisionDeformer::~CollisionDeformer() {}

void* CollisionDeformer::creator()
{
    return new CollisionDeformer();
}

MStatus CollisionDeformer::initialize()
{
    MFnNumericAttribute nAttr;
    aBounciness = nAttr.create("bounciness", "b", MFnNumericData::kFloat, 1.0);
    nAttr.setKeyable(true);
    addAttribute(aBounciness);
    attributeAffects(aBounciness, outputGeom);

    aFriction = nAttr.create("friction", "f", MFnNumericData::kFloat, 0.5);
    nAttr.setKeyable(true);
    addAttribute(aFriction);
    attributeAffects(aFriction, outputGeom);

    return MS::kSuccess;
}

MStatus CollisionDeformer::deform(MDataBlock& data, MItGeometry& itGeo, const MMatrix& localToWorldMatrix, unsigned int geomIndex)
{
    MStatus status;

    float bounciness = data.inputValue(aBounciness).asFloat();
    float friction = data.inputValue(aFriction).asFloat();

    MFnMesh fnMesh(inputGeom, &status);
    CHECK_MSTATUS_AND_RETURN_IT(status);

    for (; !itGeo.isDone(); itGeo.next()) {
        MPoint point = it
        // perform collision calculations here
        // use bounciness and friction attributes to control the behavior
        // of the collision

        fnMesh.setPoint(itGeo.index(), point, MSpace::kObject);
    }

    return MS::kSuccess;
}
